src/output/plugins/visualization/VisualizationServer.cxx

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright The Music Player Daemon Project
+
+#include "VisualizationServer.hxx"
+
+#include "Log.hxx"
+#include "lib/fmt/ThreadIdFormatter.hxx"
+#include "config/Block.hxx"
+#include "util/Domain.hxx"
+
+using std::make_unique, std::move;
+
+const Domain vis_server_domain("vis_server");
+
+Visualization::VisualizationServer::VisualizationServer(
+	EventLoop &event_loop,
+	const char *bind_to_address,
+	uint16_t port,
+	size_t max_clients_in,
+	const SoundAnalysisParameters &params_in)
+: ServerSocket(event_loop),
+  max_clients(max_clients_in),
+  reaper(event_loop, BIND_THIS_METHOD(ReapClients)),
+  sound_params(params_in)
+{
+	FmtInfo(vis_server_domain, "VisualizationServer::VisualizationServer("
+		"{}:{}, {} clients maximum)", bind_to_address, port,
+		max_clients);
+
+	ServerSocketAddGeneric(*this, bind_to_address, port);
+}
+
+void
+Visualization::VisualizationServer::ReapClients() noexcept
+{
+	FmtNotice(vis_server_domain, "VisualizationServer::ReapClients({}, "
+			  "{} clients)", std::this_thread::get_id(), clients.size());
+
+	for (auto p0 = clients.begin(), p1 = clients.end(); p0 != p1; ) {
+		auto p = p0++;
+		if (p->IsClosed()) {
+			LogInfo(vis_server_domain, "Reaping closed client.");
+			clients.erase(p);
+		}
+	}
+
+	if (!clients.empty()) {
+		LogInfo(vis_server_domain, "Scheduling another reaping in 3 "
+			"seconds.");
+		reaper.Schedule(std::chrono::seconds(3));
```

Comment:
This periodically calls ReapClients, all the time, while MPD runs! This is bad, because it causes unnecessary CPU wakeups. Never wake up the CPU unless you explicitly have a reason to do so.
I also don't understand the point of this whole method. Why does it exist? If a client is closed, why not delete it immediately?